### PR TITLE
Fix LRUness of ScrollAwareImageProvider

### PR DESCRIPTION
**Pull request by dnfield**
## Description

https://github.com/flutter/flutter/pull/49389 introduced a subtle bug where, if an image got precached _while you are scrolling fast_, the scrolling image provider would not update the ImageCache's LRU information.  The test for this is very highly constructed to push this, as it would be very hard to do this in a real application.  I actually discovered this while working through a solution for https://github.com/flutter/flutter/issues/49456, where this will become more important, but it's fixable by itself so I'm splitting it out.

## Related Issues

The two linked bugs above.

## Tests

I added the following tests:

A test that precaches an image while scrolling fast, and then observes that our image overwrites it in the cache (cache size = 1) once it completes, still while scrolling fast.  This test is highly constructed, and would be very difficult to achieve in a real application without writing custom image providers.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/image_provider.dart [hunk 1]
```diff
@@ -411,6 +411,17 @@ abstract class ImageProvider<T> {
   /// [ImageCache].
   @protected
   void resolveStreamForKey(ImageConfiguration configuration, ImageStream stream, T key, ImageErrorListener handleError) {
+    // This is an unusual edge case where someone has told us that they found
+    // the image we want before getting to this method. We should avoid calling
+    // load again, but still update the image cache with LRU information.
+    if (stream.completer != null) {
+      PaintingBinding.instance.imageCache.putIfAbsent(
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/image_provider.dart]: Maybe assert after this that it returned `stream.completer` as expected?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/painting/image_provider.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/image_provider.dart]: Done

**Comment by dnfield** [comment]
The tree is not red, manually checked - we're having issues with cocoon right now when it tries to check status.  Landing.

**Pull request merged**